aya/src/interface.rs

```diff
@@ -0,0 +1,168 @@
+//! Network interfaces representing kernel network interfaces.
+
+use std::ffi::CStr;
+
+use libc::if_nameindex;
+
+#[derive(Debug, Clone)]
+/// A kernel network interface.
+// if_name isn't stored because it can change in the kernel and aya won't know
+pub struct NetworkInterface {
+    pub(crate) index: i32,
+}
+
+impl NetworkInterface {
+    /// Provides a number that can be used to identify this interface on this system.
+    pub fn index(&self) -> i32 {
+        self.index
+    }
+
+    /// Extracts the interface name from the kernel.
+    pub fn name(&self) -> Result<String, std::io::Error> {
+        let mut buffer: [libc::c_char; libc::IF_NAMESIZE] = [0; libc::IF_NAMESIZE];
+        let name = unsafe {
+            // Returns null on error
+            let res = libc::if_indextoname(self.index as u32, buffer.as_mut_ptr());
+
+            if res.is_null() {
+                return Err(std::io::Error::last_os_error());
+            }
+
+            CStr::from_ptr(buffer.as_ptr())
+        };
+
+        Ok(name.to_string_lossy().to_string())
+    }
+
+    /// Provides a [Vec] of all operating system network interfaces, including virtual ones.
+    /// # Example
+    ///
+    /// ```
+    /// let interfaces_names: Vec<String> = NetworkInterface::list()
+    ///     .iter()
+    ///     .map(|interface| interface.name().unwrap())
+    ///     .collect();
+    /// ```
+    pub fn list() -> Vec<NetworkInterface> {
+        let mut list = Vec::new();
+
+        // The nameindex array is terminated by an interface with if_index == 0 and if_name == null
+        let head = unsafe { libc::if_nameindex() };
+        let mut curr = head;
+
+        while let Ok(interface) = NetworkInterface::try_from(unsafe { *curr }) {
+            list.push(interface);
+            curr = unsafe { curr.add(1) };
+        }
+
+        unsafe {
+            libc::if_freenameindex(head);
+        };
+
+        list
+    }
+}
+
+impl TryFrom<if_nameindex> for NetworkInterface {
+    type Error = ();
+
+    // Returns Err is the interface is invalid (zeroed)
+    fn try_from(value: if_nameindex) -> Result<Self, ()> {
```

Comment:
You should just inline this in the while loop where you're calling it. If you make it `TryFrom`, it becomes a public API which is not what we want.

Reply:
I made a private function `Interface::from_ifnameindex()`.